search/elastic.py

```diff
@@ -23,52 +21,74 @@
-        translated_result = copy.copy(result)
-        data = translated_result.pop("_source")
+        """
+        Any conversion from ES result syntax into our search engine syntax
+        """
+        result["data"] = result.pop("_source")
+        result["score"] = result.pop("_score")
+        return result
 
-        translated_result.update({
-            "data": data,
-            "score": translated_result["_score"]
-        })
+    def translate_agg_bucket(bucket, agg_result):
+        """
+        Any conversion from ES aggregations result into our search engine syntax
 
-        return translated_result
+        agg_result argument needs for getting total number of
+        documents per bucket
 
-    def translate_facet(result):
-        """ Any conversion from ES facet syntax into our search engine sytax """
-        terms = {term["term"]: term["count"] for term in result["terms"]}
+        :param bucket: string
+        :param agg_result: dict
+        :return: dict
+        """
+        agg_item = agg_result[bucket]
+        terms = {
+            bucket["key"]: bucket["doc_count"]
+            for bucket in agg_item["buckets"]
+        }
+        total_docs = (
+            agg_result[_get_total_doc_key(bucket)]["value"]
+            + agg_item["sum_other_doc_count"]
+            + agg_item["doc_count_error_upper_bound"]
+        )
         return {
             "terms": terms,
-            "total": result["total"],
-            "other": result["other"],
+            "total": total_docs,
+            "other": agg_item["sum_other_doc_count"],
         }
 
     results = [translate_result(hit) for hit in es_response["hits"]["hits"]]
     response = {
         "took": es_response["took"],
-        "total": es_response["hits"]["total"],
+        "total": es_response["hits"]["total"]["value"],
         "max_score": es_response["hits"]["max_score"],
         "results": results,
     }
-
-    if "facets" in es_response:
-        response["facets"] = {facet: translate_facet(es_response["facets"][facet]) for facet in es_response["facets"]}
+    if "aggregations" in es_response:
+        response["aggs"] = {
+            bucket: translate_agg_bucket(bucket, es_response["aggregations"])
+            for bucket in es_response["aggregations"]
+            if "total_" not in bucket
+        }
 
     return response
 
 
 def _get_filter_field(field_name, field_value):
-    """ Return field to apply into filter, if an array then use a range, otherwise look for a term match """
-    filter_field = None
+    """
+    Return field to apply into filter, if an array then use a range,
+    otherwise look for a term match
+    """
+    filter_field = {"term": {field_name: field_value}, }
     if isinstance(field_value, ValueRange):
         range_values = {}
         if field_value.lower:
-            range_values.update({"gte": field_value.lower_string})
+            range_values["gte"] = field_value.lower_string
```

Comment:
Why do we need to update dictionary with the elements from the another dictionary if we can just add new key value to the dict structure

Reply:
Please add an example of the efficiency of the adding element comparing to updating.

The main rule we do not change code just because we like one approach more than another. Code can be changed if it is required by the implemented feature, bug fixing, or fixing an evident error. Otherwise, we have additional strings for the review without any meaningful background that is not a good practice.

Reply:
Igor, let me help Sergey a little. I know you prefer the `update` method and you may have a reason for that I hope. But I'm on Sergey's side here. I really do not like this solution in python `foo_dict.update ({" bar ":" baz "})`

When I see this, I want to ask if a person understands how python works, how a dictionary works in python and how super inefficient it is. This is all micro-optimization, but in the end, from the sum of such optimizations, we can get a normal program, and not something slow and fat in memory.

What's the worst going on here? We create a dictionary object `{"bar": "baz"}`. Is one object being created? There is not one, unfortunately, but a whole tree of objects, because we are all python and in python, everything is an object, we need to remember this. Any object has a bunch of attributes, the attributes of which are also objects. One has only to think about `__dict__`, etc. It's a really big tree. An empty dictionary in python size `50 bytes` if I'm not mistaken. But the size was really big. When I found out about this, I was very surprised. And this is a big problem.
In context with all this, the question arises in me - why do this? Why create unnecessary objects ?!

A dictionary in python is a hashmap, and it adds and removes items by key nicely. This is its strong point. Let's use this. Let's use the strengths of python and be aware of the weaknesses.

The `update` method can be useful when we need to add another dictionary to an existing dictionary. This dictionary is the result of business logic, etc., but we do not need to create an artificial dictionary, just to use the `update` method.

Reply:
I'm not a fan of the `update` I meant a little bit different thing, that feature is feature and other code should not be changed in the feature PR it must be changed in the new refactoring PR which could be very welcome.

The change itself seems reasonable so if you insist let it be ))))